preprocessing/shift_generation.py

```diff
@@ -184,29 +205,30 @@ def get_shifts_overlapping_t(shifts, time_sets):
     return shifts_overlapping_t
 
 
-def get_off_shifts(root):
+def get_off_shifts(shifts_per_week):
 
-    events = get_start_events(root)
-
-    off_shifts = []
+    off_shifts = tuplelist()
     off_shifts_in_week = tupledict()
-    week = 0
-    off_shifts_in_week[week] = []
-
-    for i in range(len(events)):
-        for event in events[i:]:
-            duration = event - events[i]
-            if events[i] >= (week + 1) * 24 * 7:
-                week += 1
-                off_shifts_in_week[week] = []
-            if event >= (week + 1) * 24 * 7:
-                break
-            if duration > 70:
-                break
-            elif duration >= 36:
-                if (events[i], duration) not in off_shifts:
-                    off_shifts_in_week[week].append((events[i], duration))
-                    off_shifts.append((events[i], duration))
+
+    for week in shifts_per_week:
+
+        off_shifts_in_week[week] = []
+
+        if shifts_per_week[week][0][0] != (24 * 7 * week):
+            shifts_per_week[week].insert(0, (0.0, 0.0))
+        if shifts_per_week[week][-1][0] != (24 * 7 * (week + 1)):
+            shifts_per_week[week].append((float(24 * 7 * (week + 1)), 0.0))
+
+        for shift in shifts_per_week[week]:
+            for s in shifts_per_week[week][shifts_per_week[week].index(shift) + 1:]:
+                end_of_work_shift = (shift[0] + shift[1])
+                duration = s[0] - end_of_work_shift
+                if duration > WEEKLY_REST_DURATION[1]:
+                    break
```

Comment:
Why do you need to break here? This means that the off-period is longer than the maximum rewarded off shift?

Reply:
Yes, it is to speed up the code. If the duration exceeds the max limit of weekly off-shift duration, it should not be necessary to loop the remaining shifts, as all of them would give a too long duration.